lib/SILOptimizer/Utils/OwnershipUtils.cpp

```diff
@@ -0,0 +1,804 @@
+  // borrow.
+  return replaceAllUsesAndEraseInner(oldValue, newBorrowedValue,
+                                     ctx.eraseNotify);
+}
+
+SILBasicBlock::iterator OwnershipRAUWUtility::perform() {
+  assert(oldValue->getFunction()->hasOwnership());
+  assert(
+      OwnershipFixupContext::canFixUpOwnershipForRAUW(oldValue, newValue) &&
+      "Should have checked if can perform this operation before calling it?!");
+  // If our new value is just none, we can pass anything to do it so just RAUW
+  // and return.
+  //
+  // NOTE: This handles RAUWing with undef.
+  if (newValue.getOwnershipKind() == OwnershipKind::None)
+    return replaceAllUsesAndEraseInner(oldValue, newValue, ctx.eraseNotify);
+  assert(SILValue(oldValue).getOwnershipKind() != OwnershipKind::None);
+
+  switch (SILValue(oldValue).getOwnershipKind()) {
+  case OwnershipKind::None:
+    // If our old value was none and our new value is not, we need to do
+    // something more complex that we do not support yet, so bail. We should
+    // have not called this function in such a case.
+    llvm_unreachable("Should have been handled elsewhere");
+  case OwnershipKind::Any:
+    llvm_unreachable("Invalid for values");
+  case OwnershipKind::Guaranteed: {
+    return handleGuaranteed();
+  }
+  case OwnershipKind::Owned: {
+    // If we have an owned value that we want to replace with a value with any
+    // other non-None ownership, we need to copy the other value for a
+    // lifetimeEnding RAUW, then RAUW the value, and insert a destroy_value on
+    // the original value.
+    auto extender = getLifetimeExtender();
+    SILValue copy =
+        extender.copyAndExtendForLifetimeEndingRAUW(newValue, oldValue);
+    cleanupOperandsBeforeDeletion(oldValue, ctx.newInstNotify);
+    auto result = replaceAllUsesAndEraseInner(oldValue, copy, ctx.eraseNotify);
+    return result;
+  }
+  case OwnershipKind::Unowned: {
+    return handleUnowned();
+  }
+  }
+  llvm_unreachable("Covered switch isn't covered?!");
+}
+
+//===----------------------------------------------------------------------===//
+//                          Ownership Fixup Context
+//===----------------------------------------------------------------------===//
+//
+// Top level entry points to RAUW code.
+//
+bool OwnershipFixupContext::canFixUpOwnershipForRAUW(
+    SingleValueInstruction *oldValue, SILValue newValue) {
+  auto newOwnershipKind = newValue.getOwnershipKind();
+
+  // If our new kind is ValueOwnershipKind::None, then we are fine. We
+  // trivially support that. This check also ensures that we can always
+  // replace any value with a ValueOwnershipKind::None value.
+  if (newOwnershipKind == OwnershipKind::None)
+    return true;
+
+  // If we are in Raw SIL, just bail at this point. We do not support
+  // ownership fixups.
+  if (oldValue->getModule().getStage() == SILStage::Raw)
+    return false;
+
+  // In constrast, if our old ownership kind is ValueOwnershipKind::None and
```

Comment:
These comments are hard to parse because the description is the inverse of the condition being commented. You could either flip the condition in the code or delete the first sentence in this comment.

Reply:
I fixed it.